### PR TITLE
Decouple Place and Expression in ExprUseVisitor Delegate

### PR DESCRIPTION
**Pull request by null-sleep**
Modifies the `Delegate` interface to take a `place: Place` and an `expr_id: HirID`, where `expr_id` is pointed to in diagnostics.

Issue: https://github.com/rust-lang/project-rfc-2229/issues/20

Example:

```
let [a, ...]      =     arr;
^^Expr1^^ = ^Expr2^
```

Here `arr` is moved because of the binding created in Expr1. However, in diagnostics we point to Expr1 with the message `arr` was moved, which can be confusing. With this change, the diagnostics will point to Expr2 for patterns.

We have tried to previously achieve something similar before https://github.com/rust-lang/rust/pull/75933/files#diff-1f18144176a6e44f905262991bed987098348cdbd688ac76f8c915ec91b7fd58R316-R319

---
This change is [Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/27)

### CODE REVIEW

File: src/test/ui/borrowck/borrowck-closures-slice-patterns.stderr [hunk 1]
```diff
@@ -75,7 +75,7 @@ LL | fn arr_box_by_move(x: Box<[String; 3]>) {
 LL |     let f = || {
    |             -- value moved into closure here
 LL |         let [y, z @ ..] = *x;
-   |                            - variable moved due to use in closure
+   |                           -- variable moved due to use in closure
```
Comment (roxelo) [inline src/test/ui/borrowck/borrowck-closures-slice-patterns.stderr:78]: Is this the only test case impacted? From my understanding of the changes, it seems like at least new test cases should be introduced to confirm the behavior is now different (if none of the other one are currently impacted).

Comment (arora-aman) [reply src/test/ui/borrowck/borrowck-closures-slice-patterns.stderr:78]: this change goes in before the new behavior

Comment (arora-aman) [reply src/test/ui/borrowck/borrowck-closures-slice-patterns.stderr:78]: Also the new behavior doesn't affect things yet

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 2]
```diff
@@ -278,12 +278,13 @@ struct InferBorrowKind<'a, 'tcx> {
 impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
     fn adjust_upvar_borrow_kind_for_consume(
         &mut self,
-        place_with_id: &PlaceWithHirId<'tcx>,
+        place: &PlaceWithHirId<'tcx>,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/check/upvar.rs]:
```suggestion
        place_with_id: &PlaceWithHirId<'tcx>,
```

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 3]
```diff
@@ -343,10 +337,17 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
     }
 
     /// Indicates that `place_with_id` is being directly mutated (e.g., assigned
-    /// to). If the place is based on a by-ref upvar, this implies that
+    ///  to). If the place is based on a by-ref upvar, this implies that
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/check/upvar.rs]: extra space

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 4]
```diff
@@ -500,29 +504,38 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
 }
 
 impl<'a, 'tcx> euv::Delegate<'tcx> for InferBorrowKind<'a, 'tcx> {
-    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, mode: euv::ConsumeMode) {
-        debug!("consume(place_with_id={:?},mode={:?})", place_with_id, mode);
-        self.adjust_upvar_borrow_kind_for_consume(place_with_id, mode);
+    fn consume(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/check/upvar.rs]: `place_with_id`

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 5]
```diff
@@ -500,29 +504,38 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
...
+    ) {
+        debug!("consume(place={:?}, diag_expr_id={:?}, mode={:?})", place, diag_expr_id, mode);
+        self.adjust_upvar_borrow_kind_for_consume(&place, diag_expr_id, mode);
     }
 
-    fn borrow(&mut self, place_with_id: &PlaceWithHirId<'tcx>, bk: ty::BorrowKind) {
-        debug!("borrow(place_with_id={:?}, bk={:?})", place_with_id, bk);
+    fn borrow(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/check/upvar.rs]: `place_with_id`

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 6]
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
...
-    // on mode.
-    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, mode: ConsumeMode);
+    // on mode. Where `diag_expr_id` is the id used for diagnostics for `place`.
+    //
+    // The reson for a separate expr_id for diagonostics is to support cases
+    // like `let pat = upvar`, in such scenarios reporting the pattern (lhs)
+    // looks confusing. Instead we prefer to report the discriminant (rhs)
+    fn consume(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]: `place_with_id`

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 7]
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
...
-    // The path at `place_with_id` is being assigned to.
-    fn mutate(&mut self, assignee_place: &PlaceWithHirId<'tcx>);
+    // `diag_expr_id` is the id used for diagnostics for `place`.
+    //
+    // The reson for a separate expr_id for diagonostics is to support cases
+    // like `let pat = upvar`, in such scenarios reporting the pattern (lhs)
+    // looks confusing. Instead we prefer to report the discriminant (rhs)
+    fn borrow(
+        &mut self,
+        place: &PlaceWithHirId<'tcx>,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]: `place_with_id`

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 8]
```diff
@@ -116,11 +140,11 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
         self.mc.tcx()
     }
 
-    fn delegate_consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>) {
-        debug!("delegate_consume(place_with_id={:?})", place_with_id);
+    fn delegate_consume(&mut self, place: &PlaceWithHirId<'tcx>, diag_expr_id: hir::HirId) {
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]: `place_with_id`

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 9]
```diff
@@ -590,11 +625,10 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
 fn copy_or_move<'a, 'tcx>(
     mc: &mc::MemCategorizationContext<'a, 'tcx>,
     place_with_id: &PlaceWithHirId<'tcx>,
+    diag_expr_id: hir::HirId,
 ) -> ConsumeMode {
-    if !mc.type_is_copy_modulo_regions(
-        place_with_id.place.ty(),
-        mc.tcx().hir().span(place_with_id.hir_id),
-    ) {
+    if !mc.type_is_copy_modulo_regions(place_with_id.place.ty(), mc.tcx().hir().span(diag_expr_id))
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]: Honestly not sure if we should be using `diag_expr_id` or `place_with_id.hir_id`

**Review by arora-aman — CHANGES_REQUESTED** [review]

File: compiler/rustc_typeck/src/check/upvar.rs (hunk 2 above)
Comment (null-sleep) [reply compiler/rustc_typeck/src/check/upvar.rs]: 👍 Missed a few, fixing

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 10]
```diff
@@ -476,13 +500,17 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
 
         match *autoref {
             adjustment::AutoBorrow::Ref(_, m) => {
-                self.delegate.borrow(base_place, ty::BorrowKind::from_mutbl(m.into()));
+                self.delegate.borrow(
+                    &base_place,
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]:
```suggestion
                    base_place,
```

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 11]
```diff
@@ -476,13 +500,17 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
...
+                    base_place.hir_id,
+                    ty::BorrowKind::from_mutbl(m.into()),
+                );
             }
 
             adjustment::AutoBorrow::RawPtr(m) => {
                 debug!("walk_autoref: expr.hir_id={} base_place={:?}", expr.hir_id, base_place);
 
-                self.delegate.borrow(base_place, ty::BorrowKind::from_mutbl(m));
+                self.delegate.borrow(&base_place, base_place.hir_id, ty::BorrowKind::from_mutbl(m));
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]:
```suggestion
                self.delegate.borrow(base_place, base_place.hir_id, ty::BorrowKind::from_mutbl(m));
```

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 12]
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                     // binding being produced.
                     let def = Res::Local(canonical_id);
                     if let Ok(ref binding_place) = mc.cat_res(pat.hir_id, pat.span, pat_ty, def) {
-                        delegate.mutate(binding_place);
+                        delegate.mutate(&binding_place, binding_place.hir_id);
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]: [](url)
```suggestion
                        delegate.mutate(binding_place, binding_place.hir_id);
```

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 13]
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
...
 
                     // It is also a borrow or copy/move of the value being matched.
+                    // In a cases of pattern like `let pat = upvar`, don't use the span
+                    // of the pattern, as this just looks confusing, instead use the span
+                    // of the discriminant.
                     match bm {
                         ty::BindByReference(m) => {
                             let bk = ty::BorrowKind::from_mutbl(m);
-                            delegate.borrow(place, bk);
+                            delegate.borrow(&place, discr_place.hir_id, bk);
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]:
```suggestion
                            delegate.borrow(place, discr_place.hir_id, bk);
```

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 14]
```diff
@@ -525,19 +553,22 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
...
                         }
                         ty::BindByValue(..) => {
-                            let mode = copy_or_move(mc, place);
+                            let mode = copy_or_move(mc, &place);
                             debug!("walk_pat binding consuming pat");
-                            delegate.consume(place, mode);
+                            delegate.consume(&place, discr_place.hir_id, mode);
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]:
```suggestion
                            delegate.consume(place, discr_place.hir_id, mode);
```

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 15]
```diff
@@ -116,11 +140,11 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
         self.mc.tcx()
     }
 
-    fn delegate_consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>) {
+    fn delegate_consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, diag_expr_id: hir::HirId) {
         debug!("delegate_consume(place_with_id={:?})", place_with_id);
 
-        let mode = copy_or_move(&self.mc, place_with_id);
-        self.delegate.consume(place_with_id, mode);
+        let mode = copy_or_move(&self.mc, &place_with_id);
```
Comment (arora-aman) [inline compiler/rustc_typeck/src/expr_use_visitor.rs]:
```suggestion
        let mode = copy_or_move(&self.mc, place_with_id);
```

**Review by arora-aman — CHANGES_REQUESTED** [review]
No need to reborrow

**Pull request merged**